### PR TITLE
Add enum visitor builder

### LINKED ISSUE #1058: Enum Visitors having the same nice lambda builders as Unions
## What happened?
Expected to be able to use the lambda form of the visitor when looking at an enum, but turns out it's not generated.

## What did you want to happen?
Have the same lambda form available to us, the same way it is for union types.

### PR DESCRIPTION
**Pull request by mswintermeyer**
## Before this PR
Union types' visitor classes have a nice builder to generate them more concisely, but enum types do not

## After this PR
Addresses https://github.com/palantir/conjure-java/issues/1058

==COMMIT_MSG==
Enum types now have a visitor builder
==COMMIT_MSG==

## Possible downsides?
Implemented this just by copying the relevant methods over from UnionGenerator and adapting them so they fit the enum semantics and types. There's potential for code reuse in future, but seemed like more effort than needed right now since this functionality has been desired for over 3 years (see when the linked issue was created).

### CODE REVIEW

File: conjure-java-core/src/test/java/com/palantir/conjure/java/types/EnumTests.java
Comment (ash211) [inline conjure-java-core/src/test/java/com/palantir/conjure/java/types/EnumTests.java:105]: please add test coverage for throwOnUnknown() as well

**Review by ash211 — COMMENTED** [review]
Thank you for sending!  I've often wanted this myself as a fellow user.

I don't have approval permissions on this repo, but left a comment for a missing test case that I think the approvers would request.

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/EnumGenerator.java
Comment (mpritham) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/types/EnumGenerator.java:67]: Do we want this to also be `private`?

**Review by carterkozak — APPROVED** [review]
looks great, thanks @mswintermeyer!

File: conjure-java-core/src/main/java/com/palantir/conjure/java/types/EnumGenerator.java
Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/EnumGenerator.java:67]: Good catch, `private static final` would be better, I don't think there are uses in other classes. Similarly the `UNKNOWN_TYPE_PARAM_NAME` above should also be private.

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/types/EnumGenerator.java:67]: (it's fine if this releases as is I'd be happy to approve a fixup later since this module isn't used as a library, so we don't have to worry about folks consuming internal fields)

**Pull request merged**
